rest/routes.go

```diff
@@ -630,3 +632,71 @@ func (s *Service) setServiceFlagDisabled(w http.ResponseWriter, r *http.Request)
 	resp.State = true
 	gimlet.WriteJSON(w, &resp)
 }
+
+type userCredentials struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+type userAPIKeyResponse struct {
+	Username string `json:"username"`
+	Key      string `json:"key"`
+}
+
+func (s *Service) fetchUserToken(rw http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	creds := &userCredentials{}
+	if err := gimlet.GetJSON(r.Body, creds); err != nil {
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(errors.Wrap(err, "problem reading request body")))
+		return
+	}
+
+	if creds.Username == "" {
+		gimlet.WriteJSONResponse(rw, http.StatusUnauthorized, gimlet.ErrorResponse{
+			Message:    "no username specified",
+			StatusCode: http.StatusUnauthorized,
+		})
+	}
+
+	resp := &userAPIKeyResponse{Username: creds.Username}
+
+	token, err := s.UserManager.CreateUserToken(creds.Username, creds.Password)
+	if err != nil {
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(errors.Wrap(err, "problem creating user token")))
+		return
+	}
+
+	user, err := s.UserManager.GetUserByToken(r.Context(), token)
+	if err != nil {
+		gimlet.WriteResponse(rw, gimlet.MakeJSONErrorResponder(errors.Wrap(err, "problem finding user")))
+		return
+	}
+
+	key := user.GetAPIKey()
+	if key != "" {
+		resp.Key = key
+		gimlet.WriteJSON(rw, resp)
+		return
+	}
+
+	dbuser, ok := user.(*model.DBUser)
+	if !ok {
+		gimlet.WriteJSONResponse(rw, http.StatusInternalServerError, gimlet.ErrorResponse{
+			Message:    "cannot generate key for user",
+			StatusCode: http.StatusInternalServerError,
+		})
+
+		return
+	}
+
+	k := util.RandomString()
```

Comment:
Why not create the api key in model/users.go? Similar to how we create cookie tokens there. I think we should keep user related code there.